test/tap/tests/mysql-protocol_compression_level-t.cpp

```diff
@@ -0,0 +1,184 @@
+			mysql_errno(proxysql_admin), mysql_error(proxysql_admin)
+		);
+		return compression_level;
+	}
+
+	MYSQL_RES* res =  mysql_store_result(proxysql_admin);
+	if (res != nullptr) {
+		int num_rows = mysql_num_rows(res);
+		MYSQL_ROW row = mysql_fetch_row(res);
+
+		if (num_rows && row != nullptr) {
+			char* endptr = nullptr;
+			compression_level = strtol(row[1], &endptr, SERVERS_COUNT);
+		}
+	}
+	mysql_free_result(res);
+	return compression_level;
+}
+
+MYSQL* initilize_mysql_connection(char* host, char* username, char* password, int port, bool compression) {
+	MYSQL* mysql = mysql_init(NULL);
+	if (!mysql)
+		return nullptr;
+
+	fprintf(stderr, "MySQL connection details: %s %s %d\n", username, password, port);
+	if (!mysql_real_connect(mysql, host, username, password, NULL, port, NULL, 0)) {
+	    fprintf(stderr, "Failed to connect to database: Error: %s\n",
+	              mysql_error(mysql));
+		mysql_close(mysql);
+		return nullptr;
+	}
+	if (compression) {
+		if (mysql_options(mysql, MYSQL_OPT_COMPRESS, nullptr) != 0) {
+			fprintf(stderr, "Failed to set mysql compression option: Error: %s\n",
+					mysql_error(mysql));
+			mysql_close(mysql);
+			return nullptr;
+		}
+	}
+	return mysql;
+}
+
+int main(int argc, char** argv) {
+
+	CommandLine cl;
+
+	if(cl.getEnv())
+		return exit_status();
+
+	plan(5);
+
+	// ProxySQL connection without compression
+	MYSQL* proxysql = initilize_mysql_connection(cl.host, cl.username, cl.password, cl.port, false);
+	if (!proxysql) {
+		return exit_status();
+	}
+
+	// ProxySQL connection with compression
+	MYSQL* proxysql_compression = initilize_mysql_connection(cl.host, cl.username, cl.password, cl.port, true);
+	if (!proxysql_compression) {
+		return exit_status();
+	}
+
+	// MySQL connection with compression
+	MYSQL* mysql_compression = initilize_mysql_connection(cl.host, cl.username, cl.password, cl.mysql_port, true);
+	if (!mysql_compression) {
+		return exit_status();
+	}
+
+	// ProxySQL admin connection
+	MYSQL* proxysql_admin = initilize_mysql_connection(cl.host, cl.admin_username, cl.admin_password, cl.admin_port, false);
+	if (!proxysql_admin) {
+		return exit_status();
+	}
+
+	MYSQL_QUERY(proxysql, "CREATE DATABASE IF NOT EXISTS test");
+	MYSQL_QUERY(proxysql, "DROP TABLE IF EXISTS test.sbtest1");
+
+	mysql_query(proxysql, "CREATE TABLE IF NOT EXISTS test.sbtest1 (id INT UNSIGNED NOT NULL AUTO_INCREMENT, k INT UNSIGNED NOT NULL DEFAULT 0, c CHAR(120) NOT NULL DEFAULT '', pad CHAR(60) NOT NULL DEFAULT '', PRIMARY KEY (id), KEY k_1 (k));");
+
+	MYSQL_QUERY(proxysql, "USE test");
+
+	for (int i = 0; i < 100; i++) {
+		MYSQL_QUERY(proxysql, "INSERT INTO sbtest1 (k, c, pad) SELECT FLOOR(RAND() * 10000), REPEAT('a', 120), REPEAT('b', 60) FROM information_schema.tables LIMIT 1000;");
+	}
+
+	std::string query = "SELECT t1.id id1, t1.k k1, t1.c c1, t1.pad pad1, t2.id id2, t2.k k2, t2.c c2, t2.pad pad2 FROM test.sbtest1 t1 JOIN test.sbtest1 t2 LIMIT 90000000";
```

Comment:
For performing `SELECT` queries right after `INSERT` queries, we should either:

- Disable potential query rules that can lead to query redirection to replicas.
- Insert query annotation in the `SELECT` themselves that target the `primary` hostgroup.

Otherwise slow replication could lead to errors or not performing the testing as it was intended, since all the required data may not be in the target `replica` server.

Reply:
Fixed